src/profiler.cpp

```diff
@@ -1641,91 +1641,99 @@ void Profiler::dumpText(Writer& out, Arguments& args) {
+    size_t engine_type_strindex;
+    size_t engine_units_strindex;
+    size_t count_strindex;
+};
+
+static void recordSampleType(ProtoBuffer& otlp_buffer, size_t type_strindex, size_t unit_strindex) {
     using namespace Otlp;
     protobuf_mark_t sample_type_mark = otlp_buffer.startMessage(Profile::sample_type, 1);
-    otlp_buffer.field(ValueType::type_strindex, strings.indexOf(type));
-    otlp_buffer.field(ValueType::unit_strindex, strings.indexOf(units));
-    otlp_buffer.field(ValueType::aggregation_temporality, AggregationTemporality::cumulative);
+    otlp_buffer.field(ValueType::type_strindex, type_strindex);
+    otlp_buffer.field(ValueType::unit_strindex, unit_strindex);
     otlp_buffer.commitMessage(sample_type_mark);
 }
 
-void Profiler::dumpOtlp(Writer& out, Arguments& args) {
-    using namespace Otlp;
-    ProtoBuffer otlp_buffer(OTLP_BUFFER_INITIAL_SIZE);
-    Index strings;
-    Index functions;
-    // Eventually this is going to be Index<Attribute>, for now we keep it simple
-    Index thread_names;
-
-    protobuf_mark_t resource_profiles_mark = otlp_buffer.startMessage(ProfilesData::resource_profiles);
-    protobuf_mark_t scope_profiles_mark = otlp_buffer.startMessage(ResourceProfiles::scope_profiles);
-    protobuf_mark_t profile_mark = otlp_buffer.startMessage(ScopeProfiles::profiles);
-
-    u64 time_nanos = _start_time * 1000ULL;
-    u64 duration_nanos = (OS::micros() - _start_time) * 1000ULL;
-
-    otlp_buffer.field(Profile::time_nanos, time_nanos);
-    otlp_buffer.field(Profile::duration_nanos, duration_nanos);
-
-    recordSampleType(otlp_buffer, strings, _engine->type(), "count");
-    recordSampleType(otlp_buffer, strings, _engine->type(), _engine->units());
-
-    std::vector<CallTraceSample*> call_trace_samples;
-    _call_trace_storage.collectSamples(call_trace_samples);
+struct SampleInfo {
+    u64 samples;
+    u64 counter;
+    size_t thread_name_index;
+};
 
-    std::vector<size_t> location_indices;
-    location_indices.reserve(call_trace_samples.size());
+std::vector<SampleInfo> recordStacks(ProtoBuffer& otlp_buffer, FrameName& fn, Index& thread_names_index, Index& functions,
+                                     const std::vector<CallTraceSample*>& call_trace_samples) {
+    using namespace Otlp;
 
-    FrameName fn(args, args._style & ~STYLE_ANNOTATE, _epoch, _thread_names_lock, _thread_names);
-    size_t frames_seen = 0;
+    std::vector<SampleInfo> samples_info;
     for (const auto& cts : call_trace_samples) {
         CallTrace* trace = cts->acquireTrace();
         if (trace == NULL || excludeTrace(&fn, trace) || cts->samples == 0) continue;
 
-        protobuf_mark_t sample_mark = otlp_buffer.startMessage(Profile::sample, 1);
-        otlp_buffer.field(Sample::locations_start_index, frames_seen);
-        otlp_buffer.field(Sample::locations_length, trace->num_frames);
+        samples_info.push_back(SampleInfo{cts->samples, cts->counter, 0 /* thread_name_index */});
```

Comment:
nit: You can push SampleInfo after you get thread_name_index to avoid rewriting it in place.